### PR TITLE
Implement a single point for "Check timeout expired" test result

### PR DESCRIPTION
**Pull request by Felixoid**
### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Make events like [timeouts](https://play.clickhouse.com/play?user=play#U0VMRUNUICogRlJPTSBjaGVja3MgV0hFUkUgdGVzdF9uYW1lID09ICdDaGVjayB0aW1lb3V0IGV4cGlyZWQnIEFORCBjaGVja19zdGFydF90aW1lIEJFVFdFRU4gdG9EYXRlKCcyMDI0LTA1LTEwJykgQU5EIHRvRGF0ZSgnMjAyNC0wNS0xNScp) visible in CI DB

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 14e14d61b0ea1aa6eaa1d2292e5ecef744f90772 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63982/14e14d61b0ea1aa6eaa1d2292e5ecef744f90772/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by maxknv** [comment]
I think we should not backport such prs

**Comment by Felixoid** [comment]
Why? Doesn't having `Check timeout expired` for release PR look reasonable?

File: tests/ci/ci.py [hunk 1]
```diff
@@ -1873,26 +1875,27 @@ def _run_test(job_name: str, run_command: str) -> int:
...
-    )
+    stopwatch = Stopwatch()
+    job_log = Path(TEMP_PATH) / "job_log.txt"
+    with TeePopen(run_command, job_log, env, timeout) as process:
+        retcode = process.wait()
+        if retcode != 0:
+            print(f"Run action failed for: [{job_name}] with exit code [{retcode}]")
+            if timeout and process.timeout_exceeded:
+                print(f"Timeout {timeout} exceeded, dumping the job report")
+                JobReport(
```
Comment (maxknv) [inline tests/ci/ci.py:1888]: Currently all CI job scripts:
1. do the work
2. create JobReport for further processing by ci.py script

You added a new case where it's not true any more, ci job may be terminated and job report will be created from ci.py. Is there any strong reason to create this corner case? Why can't we do the same from the job scripts?

File: tests/ci/ci.py [hunk 2]
```diff
@@ -1873,26 +1875,27 @@ def _run_test(job_name: str, run_command: str) -> int:
...
+                    status=ERROR,
+                    description=f"Timeout {timeout} exceeded",
+                    test_results=[TestResult.create_check_timeout_expired(timeout)],
+                    start_time=stopwatch.start_time_str,
+                    duration=stopwatch.duration_seconds,
+                    additional_files=[job_log],
+                )
```
Comment (maxknv) [inline tests/ci/ci.py:1895]: to actually store a job report you need call dump() on its object

**Review by maxknv — COMMENTED** [review]
What's the goal?do you want to see "timeout expired" in commit status?

**Comment by maxknv** [comment]
> Why? Doesn't having `Check timeout expired` for release PRs look reasonable?
It's reasonable, but it's just my general feeling that it might be risky to spread changes like this across all release branches. If  something goes wrong might be hard to hot fix it everywhere

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: The reason is that here is a single place to handle timeouts.

We need not the statuses, but [test_name=Check timeout expired](https://play.clickhouse.com/play?user=play#U0VMRUNUICogRlJPTSBjaGVja3MgV0hFUkUgdGVzdF9uYW1lID09ICdDaGVjayB0aW1lb3V0IGV4cGlyZWQnIEFORCBjaGVja19zdGFydF90aW1lIEJFVFdFRU4gdG9EYXRlKCcyMDI0LTA1LTEwJykgQU5EIHRvRGF0ZSgnMjAyNC0wNS0xNScp) in the CI DB.

Now, every script having a timeout in ci_config implements the process around it like animals in a zoo. One process env, another argument. And the third one is in damn docker.

We need to have a single place for it. Maybe there's a better place where to do it. But we need to do it only once.

File: tests/ci/functional_test_check.py [hunk 3]
```diff
@@ -224,12 +221,7 @@ def main():
     assert (
         check_name
     ), "Check name must be provided as an input arg or in CHECK_NAME env"
-    kill_timeout = args.kill_timeout or int(os.getenv("KILL_TIMEOUT", "0"))
```
Comment (maxknv) [inline tests/ci/functional_test_check.py:227]: bugfix_validate jobs also calls this script with timeout argument.

Comment (Felixoid) [reply tests/ci/functional_test_check.py:227]: I went across all the jobs, there's no such args

```python
JobNames.BUGFIX_VALIDATE: TestConfig(
    "",
    # we run this check by label - no digest required
    job_config=JobConfig(
        run_by_label="pr-bugfix", run_command="bugfix_validate_check.py"
    ),
),
```

Comment (Felixoid) [reply tests/ci/functional_test_check.py:227]: I see, in the script. Thanks

Comment (Felixoid) [reply tests/ci/functional_test_check.py:227]: See, that's actually what I mean. It's a fourth place where timeout is set as somebody wanted..

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: https://play.clickhouse.com/play?user=play#U0VMRUNUICogRlJPTSBjaGVja3MgV0hFUkUgdGVzdF9uYW1lID09ICdDaGVjayB0aW1lb3V0IGV4cGlyZWQnIEFORCB0b0RhdGUoY2hlY2tfc3RhcnRfdGltZSkgPT0gdG9EYXRlKCcyMDI0LTA1LTE2Jyk=

That's what we need, + https://s3.amazonaws.com/clickhouse-test-reports/63982/03a1487a30cc52ae0aebde0fcb7ed48c7dc1b348/style_check.html

Comment (maxknv) [reply tests/ci/ci.py:1888]: ok. I agree, it makes sense to control execution time from ci.py

I see we have timeout inside stress_test and fast_test - it should be aligned with this new approach

What do you think about reporting timeout not as special/virtual _test_name_ but as a _check_status_ ?

**Comment by Felixoid** [comment]
> might be risky to spread changes like this across all release branches

In that case, we'll fix it in master and backport again.

Having up to 6 different CI, one modern and 5 deprecated, is a thing difficult to manage.

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: Check status is **mostly** [equal](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXR1cyBGUk9NIGNoZWNrcyBXSEVSRSAgdG9EYXRlKGNoZWNrX3N0YXJ0X3RpbWUpID09IHRvRGF0ZSgnMjAyNC0wNS0xNicpIEdST1VQIEJZIEFMTA==) to valid states of commit status.

I'd rather fix [install_check.py](https://play.clickhouse.com/play?user=play#U0VMRUNUICogRlJPTSBjaGVja3MgV0hFUkUgY2hlY2tfc3RhdHVzID09ICdPSycgQU5EIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+PSB0b0RhdGUoJzIwMjQtMDUtMTAnKSBHUk9VUCBCWSBBTEwgTElNSVQgMTA=) and have this completely equal.

**Comment by maxknv** [comment]
maybe we can consider adding default timeout for all job as well, like 2h with fixing all jobs that do not fit (for another PR)

File: tests/ci/ci.py (hunk 1 above)
Comment (maxknv) [reply tests/ci/ci.py:1888]: check_status is just a right place for that, even if github does not have such value

Comment (Felixoid) [reply tests/ci/ci.py:1888]: But what will be the test_name?

File: tests/ci/ci.py [hunk 4]
```diff
@@ -1873,26 +1875,27 @@ def _run_test(job_name: str, run_command: str) -> int:
...
+    stopwatch = Stopwatch()
+    job_log = Path(TEMP_PATH) / "job_log.txt"
+    with TeePopen(run_command, job_log, env, timeout) as process:
+        retcode = process.wait()
+        if retcode != 0:
+            print(f"Run action failed for: [{job_name}] with exit code [{retcode}]")
+            if timeout and process.timeout_exceeded:
+                print(f"Timeout {timeout} exceeded, dumping the job report")
+                JobReport(
+                    status=ERROR,
```
[the remark this answers is not in the document]
Comment (Felixoid) [reply tests/ci/ci.py:1889]: We don't

```python
def is_already_finished_by_status(self) -> bool:
    # currently we agree even for failed statuses
    for status in self.statuses:
        if self.check_name in status.context and status.state in (
            SUCCESS,
            FAILURE,
        ):
            return True
    return False
```

So FAILURE should be used here

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: I've grep for `check_status`. Not sure if I follow the idea.

Can you explain?

**Comment by Felixoid** [comment]
Another place to touch is Clickbench https://github.com/ClickHouse/ClickHouse/pull/57712

File: tests/ci/ci.py (hunk 1 above)
Comment (maxknv) [reply tests/ci/ci.py:1888]: no test case - no test_name, right?

File: tests/ci/ci.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxknv) [reply tests/ci/ci.py]: if it's not "ERROR" then what it is? I think it's either ERROR or some new value, for instance TIMEOUT

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: I don't see a point. The state is used in GitHub commit status. It's a set of 4 valid strings.

**Comment by Felixoid** [comment]
I'll add timeout for ClickBench around [600 seconds](https://play.clickhouse.com/dashboard#N4IgFg9gzgLiBc4YwA5XgegygNgQwE8A6AYxwEsSBrSAVygFNSIBbEAGhHoYCcERchDiACOtXuQZQEAbVAxyMHA34AhcgDs8PAgAIARrXI4AJtM5jeBfgGUAogBk7AYQAqAHQ26YEGzG0wAPIAZgCSGjC8AG54OAAUsAEMJgD6eDDsuqEAcq52AEoAagCCDrrAPBC0GiYAElU86ACq4TAAzABMAL4AVABsAAy6ALI5TXkAlPDwOa66xTa6dDzsnrpaLAy6eFAGRqYpGwyrXnhRAOZxJKy4DJGp6boAtLqJPPdpMBPzi2fnKQpNp4AGL5QLDXQmBjBPC0HAwIgAdwgPCowRwEERKQAVhB9FBPAB1WoFOy6NZvD6PVR2VyEux2bLlYKVFjwPw8TTnLrzbIAEXKPnZME5Gm5vL5FP8MHougAvLoAOTXFi3e6KiVra4aMj0cgQLwKxVQWgkEhSKAa4r8tY8BgoaCKFF6I3OCjUercDBuyhUT2MRVra0Co5ZByhADSZMVAFIyL79AwdWAY4GNABxMFNAAKulUAE0lg1MoZjKkjp5Avk+QU84Xlpk-gDyJtdDWbM4S-ty3hNiAugBdTgobS96TwUCVap1BrjkAdYQs1j8DoDDoAFieA03AwAjMIfCu15vd7unh1911OIxtCQwClLDpbI4XHMFEpjrpH3pQeDXgRYAYFgiBMHYwH0CBtDMXRiVJSEwIgqD5SVQIomiSREUVYQSHoHwWHIAAvZIEBFcQuiAA), should be good
cc @alexey-milovidov

File: tests/ci/ci.py (hunk 1 above)
Comment (Felixoid) [reply tests/ci/ci.py:1888]: Agreed to disagree and keep `Check timeout expired` to not play around non-standard status state `timeout`

**Comment by Felixoid** [comment]
Well, let's make it 900. It was very close to the end https://github.com/ClickHouse/ClickHouse/actions/runs/9129062303/job/25103116976#step:8:1743

**Pull request merged**